cli/command/stack/kubernetes/ps.go

```diff
@@ -29,78 +35,71 @@ func RunPS(dockerCli *KubeCli, options options.PS) error {
+		return fmt.Errorf("nothing found in stack: %s", stackName)
 	}
-
-	pods, err := fetchPods(namespace, podsClient)
 	if err != nil {
 		return err
 	}
-
+	pods, err := fetchPods(stackName, client.Pods(), filters)
+	if err != nil {
+		return err
+	}
 	if len(pods) == 0 {
-		return fmt.Errorf("nothing found in stack: %s", namespace)
+		return fmt.Errorf("nothing found in stack: %s", stackName)
 	}
+	return printTasks(dockerCli, options, stackName, client, pods)
+}
 
+func printTasks(dockerCli command.Cli, options options.PS, namespace string, client corev1.NodesGetter, pods []apiv1.Pod) error {
 	format := options.Format
-	if len(format) == 0 {
+	if format == "" {
 		format = task.DefaultFormat(dockerCli.ConfigFile(), options.Quiet)
 	}
-	nodeResolver := makeNodeResolver(options.NoResolve, client.Nodes())
 
 	tasks := make([]swarm.Task, len(pods))
 	for i, pod := range pods {
 		tasks[i] = podToTask(pod)
 	}
-	return print(dockerCli, namespace, tasks, pods, nodeResolver, !options.NoTrunc, options.Quiet, format)
-}
-
-type idResolver func(name string) (string, error)
-
-func print(dockerCli command.Cli, namespace string, tasks []swarm.Task, pods []apiv1.Pod, nodeResolver idResolver, trunc, quiet bool, format string) error {
 	sort.Stable(tasksBySlot(tasks))
 
 	names := map[string]string{}
 	nodes := map[string]string{}
 
-	tasksCtx := formatter.Context{
-		Output: dockerCli.Out(),
-		Format: formatter.NewTaskFormat(format, quiet),
-		Trunc:  trunc,
+	n, err := client.Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return err
 	}
-
 	for i, task := range tasks {
-		nodeValue, err := nodeResolver(pods[i].Spec.NodeName)
+		nodeValue, err := resolveNode(pods[i].Spec.NodeName, n, options.NoResolve)
 		if err != nil {
 			return err
 		}
-
 		names[task.ID] = fmt.Sprintf("%s_%s", namespace, pods[i].Name)
 		nodes[task.ID] = nodeValue
 	}
 
+	tasksCtx := formatter.Context{
+		Output: dockerCli.Out(),
+		Format: formatter.NewTaskFormat(format, options.Quiet),
+		Trunc:  !options.NoTrunc,
+	}
+
 	return formatter.TaskWrite(tasksCtx, tasks, names, nodes)
 }
 
-func makeNodeResolver(noResolve bool, nodes corev1.NodeInterface) func(string) (string, error) {
+func resolveNode(name string, nodes *apiv1.NodeList, noResolve bool) (string, error) {
 	// Here we have a name and we need to resolve its identifier. To mimic swarm behavior
-	// we need to resolve the id when noresolve is set, otherwise we return the name.
+	// we need to resolve to the id when noResolve is set, otherwise we return the name.
 	if noResolve {
-		return func(name string) (string, error) {
-			n, err := nodes.List(metav1.ListOptions{
-				FieldSelector: fields.OneTermEqualSelector(api.ObjectNameField, name).String(),
-			})
-			if err != nil {
-				return "", err
-			}
-			if len(n.Items) != 1 {
-				return "", fmt.Errorf("could not find node '%s'", name)
+		for _, node := range nodes.Items {
+			if node.Name == name {
+				return string(node.UID), nil
```

Comment:
Is the `node.UID` really what we want? In Kube, there as you've already pointed out, names are the IDs, so there isn't really an analogue to the ID-vs-hostname split of Swarmkit (in Swarmkit, you can have two nodes with the same name, but that's not possible in Kube). As such, I think it'd make more sense to just ignore `noResolve` for Kube instead of using this UID which isn't used anywhere else in the system.

Reply:
It is more about having the same UX as Swarm than being really useful... Sure we can remove the noResolve flag and simplify the code, but in that case we may break some Swarm scripts which use this flag? From the start we choose to stick to Swarm UX, even if it is not very relevant sometimes. But maybe we are wrong and we should simplify where we can, WDYT?

Reply:
Well, we can keep the flag, but I think when we translate Kube nodes to Swarm nodes, the `ID` field and the `name` field should contain the same value (i.e. the name of the node in Kube). There's no reason to expose the `node.UID` field in `docker node ls`; users will never need to know the value of `node.UID` for the purposes of `docker node ls`. It's not important that the `ID` field contains an opaque, alphanumeric string; what's important is that it's unique, and in Kube the name is unique.

The same goes for everywhere else we use `UID` in this PR.

Reply:
Thank you @wsong, but I think this a different issue then. I opened a new issue #1075 to take care of it in a follow-up so we can move forward on this PR. Does it sounds good to you?